modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala

```diff
@@ -168,42 +180,80 @@ object Enrich {
     BadRow.GenericError(processor, failure, rawPayload)
   }
 
-  def sinkChunk[F[_]: Concurrent: Parallel, B](
+  def sinkChunk[F[_]: Concurrent: Parallel, A, B](
     chunk: List[(B, Result)],
-    sinkOne: Validated[BadRow, EnrichedEvent] => F[Unit],
-    trackLatency: Option[Long] => F[Unit]
-  ): F[List[B]] =
-    chunk.parTraverse {
-      case (record, result) =>
-        val (events, collectorTstamp) = result
-        events.parTraverse_(one => sinkOne(one)).as(record) <* trackLatency(collectorTstamp)
-    }
+    env: Environment[F, A]
+  ): F[Unit] = {
+    val (bad, enriched) =
+      chunk
+        .flatMap(_._2._1)
+        .map(_.toEither)
+        .separate
+
+    val (moreBad, good) = enriched.map { e =>
+      serializeEnriched(e, env.processor, env.streamsSettings.maxRecordSize)
+        .map(bytes => (e, AttributedData(bytes, env.goodAttributes(e))))
+    }.separate
+
+    val allBad = (bad ++ moreBad).map(badRowResize(env, _))
+
+    List(
+      sinkGood(
+        good,
+        env.sinkGood,
+        env.metrics.goodCount,
+        env.metadata.observe,
+        env.sinkPii,
+        env.piiAttributes,
+        env.processor,
+        env.streamsSettings.maxRecordSize
+      ),
+      sinkBad(allBad, env.sinkBad, env.metrics.badCount),
+      env.metrics.enrichLatency(chunk.headOption.flatMap(_._2._2))
```

Comment:
I suggest update the enrichLatency only after writing good.  Otherwise we are lying slightly about the latency of data reaching the enriched stream.